From 5ae18cd61fe2aad5d1c4846514157b42bcbb6021 Mon Sep 17 00:00:00 2001
From: Dong AiSheng <aisheng.dong@nxp.com>
Date: Tue, 28 Aug 2018 15:24:49 +0800
Subject: [PATCH 4631/5242] MLK-19351-4 can: flexcan: add notes for available
 fd mode rate combination

commit  0a631c5baffdba232fcd5756e19f4a1defcbb278 from
https://source.codeaurora.org/external/imx/linux-imx.git

CAN FD can only support some specific bitrate combinations in FD mode
due to HW limitations. See below info mentioned in RM:
"To minimize errors when processing FD frames, use the same value
for FPRESDIV and PRESDIV (in CAN_CBT or CAN_CTRL1)."

Add notes in driver to let user know easily:

CAN FD supported rates combinations
* Combination 1:
*  Bitrate: 225000 375000 400000 425000 500000 875000
*  Data rate: 1000000
*
* Combination 2:
*  Bitrate:   550000 600000 625000 650000 675000 750000 775000
*             800000 850000 925000 950000 975000 1000000
*  Data rate: 1500000 2000000 2500000 3000000 3500000 4000000
*             5000000

Signed-off-by: Joakim Zhang <qiangqing.zhang@nxp.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/net/can/flexcan.c |   19 +++++++++++++++++++
 1 file changed, 19 insertions(+)

diff --git a/drivers/net/can/flexcan.c b/drivers/net/can/flexcan.c
index 563711f7..c320b38 100644
--- a/drivers/net/can/flexcan.c
+++ b/drivers/net/can/flexcan.c
@@ -272,6 +272,25 @@ struct flexcan_mb {
 #define FLEXCAN_CANFD_MBDSR_SHIFT	16
 #define FLEXCAN_CANFD_MBDSR_DEFAULT	0x6db
 
+/*
+ * NOTE:
+ * To minimize errors when processing FD frames, use the same value
+ * for FPRESDIV and PRESDIV (in CAN_CBT or CAN_CTRL1).
+ * For more details refer to the first NOTE in section CAN FD frames.
+ *
+ * CAN FD supported rates combinations
+ *
+ * Combination 1:
+ *  Bitrate: 225000 375000 400000 425000 500000 875000
+ *  Data rate: 1000000
+ *
+ * Combination 2:
+ *  Bitrate:	550000 600000 625000 650000 675000 750000 775000
+ *		800000 850000 925000 950000 975000 1000000
+ *  Data rate:	1500000 2000000 2500000 3000000 3500000 4000000
+ *		5000000
+ */
+
 /* Structure of the hardware registers */
 struct flexcan_regs {
 	u32 mcr;		/* 0x00 */
-- 
1.7.9.5

